lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -36,6 +36,69 @@ static bool hasOperandsOutsideOfBlock(Operation *op) {
   });
 }
 
+/// If _all_ of the operation's operands, which are defined outside of the
+/// operation's own block, are ConstantLike, rematerialize them inside of the
+/// operation's block. This relaxes the conservativeness incurred by the
+/// 'hasOperandsOutsideOfBlock' checks.
+static LogicalResult
+materializeConstantOperandsLocally(Operation *op, PatternRewriter &rewriter) {
+
+  // List of ConstantLike ops to clone and the operand's result index.
+  SmallVector<std::pair<Operation *, unsigned>> opsToClone;
+  SmallVector<Value> newOperands;
+  newOperands.reserve(op->getNumOperands());
+
+  for (auto operand : op->getOperands()) {
+    auto *defOp = operand.getDefiningOp();
+
+    if (!defOp || defOp->getBlock() == op->getBlock()) {
+      // Keep local operands
+      newOperands.push_back(operand);
+      continue;
+    }
+
+    if (!defOp->hasTrait<OpTrait::ConstantLike>())
+      // Optimization will remain blocked, so don't bother changing the op.
+      return failure();
+
+    for (unsigned resultIdx = 0; resultIdx < defOp->getNumResults();
```

Comment:
I don't know if there is a multi-results ConstantLike operation. I think it's reasonable to restrict this to single result.